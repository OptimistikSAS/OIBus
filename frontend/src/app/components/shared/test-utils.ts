import { Provider } from '@angular/core';
import { NgbConfig } from '@ng-bootstrap/ng-bootstrap';

const NO_ANIMATION_NGB_CONFIG: NgbConfig = { animation: false };

export const noAnimation: Provider = { provide: NgbConfig, useValue: NO_ANIMATION_NGB_CONFIG };
