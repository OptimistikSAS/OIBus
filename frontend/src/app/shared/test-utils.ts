import { Provider } from '@angular/core';
import { NgbConfig } from '@ng-bootstrap/ng-bootstrap';
import { Page } from './types';

const NO_ANIMATION_NGB_CONFIG: NgbConfig = { animation: false };

export const noAnimation: Provider = { provide: NgbConfig, useValue: NO_ANIMATION_NGB_CONFIG };

export function toPage<T>(content: Array<T>, totalElements: number = content.length, num = 0, size = Math.max(20, totalElements)): Page<T> {
  return {
    content,
    totalElements,
    number: num,
    size,
    totalPages: Math.ceil(totalElements / size)
  };
}

export function emptyPage<T>(size = 20): Page<T> {
  return {
    totalElements: 0,
    content: [] as Array<T>,
    size,
    totalPages: 0,
    number: 0
  };
}
