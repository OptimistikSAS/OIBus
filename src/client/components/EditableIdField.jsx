import React from 'react'
import PropTypes from 'prop-types'
import { Button } from 'reactstrap'
import { FaPencilAlt, FaCheck } from 'react-icons/fa'
import { OIbText } from './OIbForm'

const EditableIdField = ({ id, fromList, index, name, idChanged }) => {
  const [editing, setEditing] = React.useState(false)
  const [editingId, setEditingId] = React.useState()
  const [otherIds, setOtherIds] = React.useState()
  const [editingError, setEditingError] = React.useState()

  const handleEditName = () => {
    setEditing(true)
    setEditingId(id)
    const list = fromList.filter((e, i) => i !== index).map((e) => e[name])
    setOtherIds(list)
  }

  const isValidName = (val) => {
    let error = null
    if (otherIds.includes(val)) {
      error = 'Id already exists'
    }
    if (!error) {
      error = (((typeof val === 'string' || val instanceof String) && val !== '') ? null : 'value must not be empty')
    }
    setEditingError(error)
    return error
  }

  const onChange = (_, value) => {
    setEditingId(value)
  }

  const handleDoneEditName = () => {
    if (!editingError) {
      idChanged(index, editingId)
      setEditing(false)
    }
  }

  const style = { display: 'inline-block' }
  return (
    editing ? (
      <div style={style}>
        <OIbText
          onChange={onChange}
          value={editingId}
          valid={isValidName}
          name={name}
          inline
        />
        <Button close>
          <FaCheck
            className="oi-icon oi-icon-inline"
            onClick={handleDoneEditName}
          />
        </Button>
      </div>
    ) : (
      <div style={style}>
        {id}
        <Button close>
          <FaPencilAlt
            className="oi-icon oi-icon-inline"
            onClick={() => {
              handleEditName()
            }}
            style={{ height: 16 }}
          />
        </Button>
      </div>
    )
  )
}

EditableIdField.propTypes = {
  id: PropTypes.string.isRequired,
  fromList: PropTypes.arrayOf(Object).isRequired,
  index: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  idChanged: PropTypes.func.isRequired,
}
export default EditableIdField
