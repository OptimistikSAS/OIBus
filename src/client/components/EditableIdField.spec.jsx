import React from 'react'
import ReactDOM from 'react-dom'
import { act } from 'react-dom/test-utils'

import EditableIdField from './EditableIdField.jsx'

let container
beforeEach(() => {
  container = document.createElement('div')
  document.body.appendChild(container)
})

afterEach(() => {
  document.body.removeChild(container)
  container = null
})

describe('EditableIdField', () => {
  test('check Alert', () => {
    act(() => {
      ReactDOM.render(<EditableIdField
        id="id"
        fromList={[{ test: 'test' }]}
        index={1}
        name="name"
        idChanged={() => (1)}
      />, container)
    })
    expect(container).toMatchSnapshot()
  })
})
