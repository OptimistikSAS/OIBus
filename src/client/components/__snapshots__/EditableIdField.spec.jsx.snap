// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditableIdField check EditableIdField editing = false 1`] = `
<div>
  <div>
    id
  </div>
</div>
`;

exports[`EditableIdField check EditableIdField editing = true 1`] = `
<div>
  <div>
    <svg
      class="oi-icon mx-2"
      fill="currentColor"
      height="1em"
      id="save-icon"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 512 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
      />
    </svg>
  </div>
</div>
`;

exports[`EditableIdField check editing done 1`] = `
<div>
  <div>
    <svg
      class="oi-icon mx-2"
      fill="currentColor"
      height="1em"
      id="save-icon"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 512 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
      />
    </svg>
  </div>
</div>
`;

exports[`EditableIdField check editing done with error 1`] = `
<div>
  <div>
    <svg
      class="oi-icon mx-2"
      fill="currentColor"
      height="1em"
      id="save-icon"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 512 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
      />
    </svg>
  </div>
</div>
`;
